In [ ]:
import numpy as np
import pandas as pd
import matplotlib as plt
import sklearn
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
import keras

## QUESTION 1

In [ ]:
def sigmoid(x):
    return 1 / (1 + pow(2.71828182846, -x))

def and_gate(input1, input2):
    # Define the weights and biases for the hidden layer
    hidden_weights = [20, 20]
    hidden_bias = -30

    # Define the weights and bias for the output layer
    output_weights = [20, 20]
    output_bias = -10

    # Feedforward propagation for the hidden layer
    hidden_layer_output = sigmoid(hidden_weights[0] * input1 + hidden_weights[1] * input2 + hidden_bias)

    # Feedforward propagation for the output layer
    output_layer_output = sigmoid(output_weights[0] * hidden_layer_output + output_weights[1] * hidden_layer_output + output_bias)

    # Return the output
    return int(output_layer_output >= 0.5)

# Default input cases
default_inputs = [(0, 0), (0, 1), (1, 0), (1,1)]

# Driver test with default inputs
for input_pair in default_inputs:
    input1, input2 = input_pair
    output = and_gate(input1, input2)
    print("Input:", input_pair, "Output:", output)

# User input
input1 = int(input("Enter the first input (0 or 1): "))
input2 = int(input("Enter the second input (0 or 1): "))

output = and_gate(input1, input2)
print("Output:", output)

###############################################################################
###############################################################################
####################  ANSWERS OF QUESTIONS ####################################
### 1 ####
# I used one hidden layer

### 2 ####
## For the hidden layer:
# The Node:
#Weight 1: 20
#Weight 2: 20
#Bias: -30

#For the output layer:
# The node
#Weight 1: 20
#Weight 2: 20
# Bias: -10


Input: (0, 0) Output: 0
Input: (0, 1) Output: 0
Input: (1, 0) Output: 0
Input: (1, 1) Output: 1
Enter the first input (0 or 1): 1
Enter the second input (0 or 1): 1
Output: 1


## QUESTION 2

In [ ]:
url = 'https://raw.githubusercontent.com/CombatLawyer/COMP472/main/NNData.csv'
df = pd.read_csv(url)

In [ ]:
print(df)

    A  B  C  D  E  y
0   1  1  1  0  0  1
1   1  1  1  0  1  1
2   0  0  0  1  0  0
3   0  0  0  1  1  0
4   0  0  1  0  0  1
5   0  0  1  0  1  1
6   0  0  1  1  0  1
7   0  0  1  1  1  1
8   0  1  0  0  0  0
9   0  1  0  0  1  0
10  0  1  0  1  0  0
11  0  1  0  1  1  0
12  0  1  1  0  0  1
13  0  1  1  0  1  1
14  0  1  1  1  0  1
15  0  1  1  1  1  1
16  1  0  0  0  0  1
17  1  0  0  0  1  1
18  1  0  0  1  0  1
19  1  0  0  1  1  1
20  1  0  1  0  0  1
21  1  0  1  0  1  1
22  1  0  1  1  0  1
23  1  0  1  1  1  1


In [ ]:
url = 'https://raw.githubusercontent.com/CombatLawyer/COMP472/main/NNTestSet.csv'
test = pd.read_csv(url)

In [ ]:
print(test)

   A  B  C  D  E  y
0  1  1  0  0  0  1
1  1  1  0  0  1  1
2  0  0  0  0  0  0
3  0  0  0  0  1  0
4  1  1  0  1  0  1
5  1  1  0  1  1  1
6  1  1  1  1  0  1
7  1  1  1  1  1  1


In [ ]:
# MLP classifier initialization, solver lbfgs and relu was used because it converges faster.
mlp = MLPClassifier(solver='lbfgs', activation='relu', random_state=1, max_iter=100)

# Grid of paramaters to test which amount of layers/amount of nodes is optimal to use. Will search only these configurations.
parameter_space = {'hidden_layer_sizes': [(1), (1, 1), (1, 1, 1), (1, 1, 1, 1), (1, 1, 1, 1, 1), (2, 2), (2, 2, 2), (2, 2, 2, 2), (2, 2, 2, 2, 2), (3), (3, 3), (3, 3, 3), (3, 3, 3, 3, 3), (4), (4, 4), (4, 4, 4), (4, 4, 4, 4), (4, 4, 4, 4, 4), (5), (5, 5), (5, 5, 5), (5, 5, 5, 5), (5, 5, 5, 5, 5)]}

# Provide grid of parameters to the MLP classifier.
clf = GridSearchCV(mlp, parameter_space)

# Fit data to grid of MLPs to determine the optimal parameters.
clf.fit(df.iloc[:,:-1].values, df.iloc[:,-1:].values.reshape(len(df),))

GridSearchCV(estimator=MLPClassifier(max_iter=100, random_state=1,
                                     solver='lbfgs'),
             param_grid={'hidden_layer_sizes': [1, (1, 1), (1, 1, 1),
                                                (1, 1, 1, 1), (1, 1, 1, 1, 1),
                                                (2, 2), (2, 2, 2), (2, 2, 2, 2),
                                                (2, 2, 2, 2, 2), 3, (3, 3),
                                                (3, 3, 3), (3, 3, 3, 3, 3), 4,
                                                (4, 4), (4, 4, 4), (4, 4, 4, 4),
                                                (4, 4, 4, 4, 4), 5, (5, 5),
                                                (5, 5, 5), (5, 5, 5, 5),
                                                (5, 5, 5, 5, 5)]})

In [ ]:
# Optimal MLP predictions on the test data.
clf.predict(test.iloc[:,:-1].values)

array([1, 1, 0, 1, 1, 1, 1, 1])

In [ ]:
# Test accuracy of the optimal layers MLP.
print(f"Accuracy: {clf.score(test.iloc[:,:-1].values, test.iloc[:,-1:].values.reshape(len(test),))*100}")
print(f"Precision: {sklearn.metrics.precision_score(clf.predict(test.iloc[:,:-1].values), test.iloc[:,-1:].values.reshape(len(test),))*100}")
print(f"Recall: {sklearn.metrics.recall_score(clf.predict(test.iloc[:,:-1].values), test.iloc[:,-1:].values.reshape(len(test),))*100}")
print(f"F1 score: {sklearn.metrics.f1_score(clf.predict(test.iloc[:,:-1].values), test.iloc[:,-1:].values.reshape(len(test),))*100}")

Accuracy: 87.5
Precision: 100.0
Recall: 85.71428571428571
F1 score: 92.3076923076923


In [ ]:
# The best parameters chosen through the grid search.
print(clf.best_params_)

{'hidden_layer_sizes': (5, 5)}


In [ ]:
clf.cv_results_

{'mean_fit_time': array([0.01827874, 0.01269579, 0.00960073, 0.00814424, 0.00999093,
        0.02089849, 0.03577347, 0.13674664, 0.00863662, 0.020438  ,
        0.05799198, 0.04631777, 0.08062429, 0.00547872, 0.00671301,
        0.02241335, 0.01796045, 0.02588477, 0.01299734, 0.01507058,
        0.02482014, 0.03995261, 0.04178357]),
 'std_fit_time': array([0.00369592, 0.00430782, 0.00305906, 0.00304659, 0.00622273,
        0.02110079, 0.03066988, 0.053497  , 0.00147912, 0.01331901,
        0.02427242, 0.02710214, 0.0260538 , 0.00057423, 0.00033533,
        0.0122159 , 0.01342445, 0.00715577, 0.00108592, 0.00259595,
        0.01303395, 0.01400135, 0.01257731]),
 'mean_score_time': array([0.00244131, 0.00219526, 0.00161633, 0.00103827, 0.00294333,
        0.00121727, 0.00665331, 0.00257735, 0.00103364, 0.00291567,
        0.00141234, 0.00108972, 0.00128617, 0.00060897, 0.00067811,
        0.00076089, 0.00092359, 0.0012413 , 0.00109634, 0.00117564,
        0.00110602, 0.00114613, 0.002603

a) How many hidden layers have you used? And why?

Using the grid search the best amount of layers determined was 2 hidden layers. The hyperparameter search was done using a grid search using options from 1-5 layers with uniform layer counts to determine which neural network was the most optimal. This was done by doing cross-validation (default is 5-fold) which holds out a portion of the data, trains the model using the rest and determines the accuracy. This is repeated for each fold and then averaged. The best result was (5, 5) which had the highest mean test score. The network of (3, 3, 3, 3) and (2) was removed because it failed to converge during the grid search using the lbfgs solver.

b) How many nodes in each hidden layer and why that number of nodes in particular?

Using the grid search as a hyperparameter search the best amount of layers for that 2 hidden layer was determined as 5 nodes. This was done using the options provided (1-5) nodes in each layer with the same nodes per layer. In the same fashion the grid search was used to determine the optimal amount of layers from the options provided.

c) What is the activation function that you used and why? Did you use the same activation function in all layers? Why?

The activation function used was the ReLU (Rectified Linear Unit) function, which was specifically selected because it converges faster than the older options (six times faster than logistical and tanh). The same activation function is used in every layer for simplicity, but does not need to be the case. For specific neural networks like CNNs (Convolutional Neural Network) can use a ReLU layers for activation but then a softmax at the end to get probabilities.

d) What learning algorithm did you use to train the neural net and why?

A MLP (Multilayer Perceptrons) was used for simplicity sake. The MLP learns a function, which can then make predictions based on a suitable input. The problem was not very difficult, nor did not have a more specialized scenario that would require a more specialized neural network like CNNs (Convolutional Neural Networks) and images, or LSTMs (Long Short Term Memory Networks) which retain information over time, or GANs (Generative Adversarial Networks) which can generate data, etc.

e) Can you use one hidden layer only to solve this problem? If yes, how many nodes are you going to have in it? And why?

In [ ]:
# Linear test of the impact of the number of nodes on the test accuracy.
for i in range(1, 31):

  # Create the MLP using the specified number of nodes.
  clf = MLPClassifier(activation='relu', solver='lbfgs', random_state=1, hidden_layer_sizes=(i))

  # Fit data to the MLP.
  clf.fit(df.iloc[:,:-1].values, df.iloc[:,-1:].values.reshape(len(df),))

  # Determine accuracy using the test set.
  testAcc = clf.score(test.iloc[:,:-1].values, test.iloc[:,-1:].values.reshape(len(test),))
  print(f"A neural network with 1 layer and {i} node(s) has a test accuracy of {testAcc}")

A neural network with 1 layer and 1 node(s) has a test accuracy of 0.75
A neural network with 1 layer and 2 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 3 node(s) has a test accuracy of 0.75
A neural network with 1 layer and 4 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 5 node(s) has a test accuracy of 0.875
A neural network with 1 layer and 6 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 7 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 8 node(s) has a test accuracy of 0.75
A neural network with 1 layer and 9 node(s) has a test accuracy of 0.75
A neural network with 1 layer and 10 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 11 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 12 node(s) has a test accuracy of 1.0
A neural network with 1 layer and 13 node(s) has a test accuracy of 0.75
A neural network with 1 layer and 14 node(s) has a test accuracy o

Yes a single hidden layer can be used but the number of nodes changes the test accuracy of the neural network. Some might be over fitted giving a perfect test accuracy. Something like the 0.875 scores might be best and then the lowest number of layers would be ideal (5 nodes). However this might also be due to the small amount of data in the train/test set in general, more data would be ideal to determine what hyperparameters might be best.

f) Can we use 5 hidden layers? Is that a good idea? Justify your answer.

5 hidden layers can be used, its just likely to overfit the relationship. In general we could use hundred of layers if we wanted to, it would just be more computaitonally heavy and large for no real reason unless the relationship we are computing is difficult/has many inputs and requires the additional layers to properly classify inputs.

g) How did the neural net do in classifying the testing set? Comment on how good or bad it learned the function from the training set.

The neural network does extremely well in some cases (when the test accuracy is 100%) and in other it does less (when it dips to 75% test accuracy in the 1 node layer tests and some of the grid options have a 50% accuracy). If assuming only this is the data we have access to then we can simply find a set of hyperparameters that yeilds 100% accuracy, but this is mostly due to the small size of data we have. In the aforementioned cases, we might sometime prefer a say 87.5% accuracy nueral network as it performs consistently well, but also allows some differences that it can potentially allow it to be more flexible in unseen data over potentially overfitting the relationship. Other accuracy measures such as precision, recall and F-score can be used in more complex cases to evaluate model accuracy. Ideally we should have more data, which would change the ideal hyperparameters, and would allow us to determine if the model is truly accurate or is overfitting to the training data. In general, a neural network can always provide a reasonable learning function from the training set, the size of the nerwork might be larger/computationally expensive as result, which might not always be ideal.